src/transformers/generation_beam_constraints.py

```diff
@@ -202,6 +211,170 @@ def copy(self, stateful=False):
         return new_constraint
 
 
+class DisjunctiveTrie:
+    def __init__(self, nested_token_ids):
+        r"""
+        A helper class that builds a trie with the words represented in `nested_token_ids`.
+
+        For example, if `nested_token_ids==[[1,2,3], [1,2,4], [1,5,6]]`, then the trie is: 1 -> 2 -> 3
+         \ -> 4
+          -> 5 -> 6
+        """
+        self.max_height = max([len(one) for one in nested_token_ids])
+
+        root = dict()
+        for token_ids in nested_token_ids:
+            level = root
+            for token_id in token_ids:
+                if token_id not in level:
+                    level[token_id] = dict()
+                level = level[token_id]
+
+            # indicate end of node
+            level[-1] = dict()
+
+        self.trie = root
+
+    def next_tokens(self, current_seq):
+        """
+        The next possible tokens that will progress the trie, given the current sequence of tokens in `current_seq`.
+
+        -1 in next_tokens refers to end of some constraint
+        """
+        start = self.trie
+
+        for current_token in current_seq:
+            current_token = current_token if isinstance(current_token, int) else current_token.tolist()
```

Comment:
This line seems wrong.

The type of `current_seq` should be predetermined earlier IMO.
`tolist()` is actually most likely wrong, since lists are not hashable and so cannot index your trie dictionnaries.

Personnally I would vote that `current_seq` is forced to be a `List[int]` to prevent any `tolist()` creeping up (would have to be called before calling this function.).